src/systems/nonlinear_implicit_system.C

```diff
@@ -82,6 +82,10 @@ void NonlinearImplicitSystem::reinit ()
   // re-initialize the nonlinear solver interface
   nonlinear_solver->clear();
 
+  // force the solver to get a new preconditioner, in
+  // case reuse was set
+  nonlinear_solver->force_new_preconditioner();
```

Comment:
Is the clear() above not enough to do that?  That seems like a bug in itself, IMHO.

Reply:
No.  In the MOOSE workflow `clear()` gets called after every nonlinear iteration.  So I had to alter what `clear()` means over here to not get rid of the SNES instance (or else there would be no way to save the conditioner).

Reply:
> No. In the MOOSE workflow clear() gets called after every nonlinear iteration.

Wow ... we render the PetscNonlinearSolver uninitialized, and unless we're reusing the precondtioner we destroy the SNES entirely, after every nonlinear iteration?

What a mess.  At least you don't seem to be digging the hole any deeper here.